File: 秋招笔试/小红书2/02.cpp
##include <iostream>
#include <vector>
#include <numeric>

const int MAX_SCORE = 500000;

int main(){
    std::ios::sync_with_stdio(false);
    std::cin.tie(NULL);


    int n, m;
    std::cin >> n >> m;
    std::vector<int> freq(MAX_SCORE+1, 0);
    for (int i = 0; i < n; i++) {
        int score;
        std::cin >> score;
        freq[score]++;
    }

    std::vector<int> divisors_count(MAX_SCORE+1, 0);
    for (int k = 1; k <= MAX_SCORE; ++k) {
        if (freq[k] > 0) {
            for (int j = k; j <= MAX_SCORE; j += k) {
                divisors_count[j] += freq[k];
            }
        }
    }

    std::vector<int> multiples_count(MAX_SCORE+1, 0);
    for (int k = 1; k <= MAX_SCORE; ++k) {
        for (int j = k; j <= MAX_SCORE; j += k) {
            multiples_count[k] += freq[j];
        }
    }

   for (int q_idx = 0; q_idx < m; ++q_idx) {
        int x;
        std::cin >> x;
        int current_compatibles = multiples_count[x] + divisors_count[x];
        if (freq[x] > 0) {
            current_compatibles -= freq[x];
        }
        std::cout << current_compatibles << std::endl;
   }

   return 0;
}

File: 秋招笔试/小红书2/02.py
"""
@Time: 2025/8/24 19:24
@Author: yanzx
@Description: 


5 3
1 2 2 5 6
4
2
1

=>> 3 4 5

"""

MAX_SCORE = 5 * 10 ** 5

def solve():
    n, m = map(int, input().split())
    a =  list(map(int, input().split()))
    freq = [0] * (MAX_SCORE + 1)
    for score in a:
        freq[score] += 1
    divisors_count = [0] * (MAX_SCORE + 1)
    for k in range(1, MAX_SCORE + 1):
        if freq[k] > 0:
            for j in range(k, MAX_SCORE + 1, k):
                divisors_count[j] += freq[k]
    muliples_count = [0] * (MAX_SCORE + 1)
    for k in range(1, MAX_SCORE + 1):
        for j in range(k, MAX_SCORE + 1, k):
            muliples_count[k] += freq[j]

    results = []
    for _ in range(m):
        x = int(input())
        cur_compatiples = muliples_count[x] + divisors_count[x]
        if freq[x] > 0:
            cur_compatiples -= freq[x]
        results.append(cur_compatiples)
    
    for res in results:
        print(res)

solve()
        
    


